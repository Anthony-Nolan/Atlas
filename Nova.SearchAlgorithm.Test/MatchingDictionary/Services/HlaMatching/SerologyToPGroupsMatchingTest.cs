using System.Collections.Generic;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using Nova.SearchAlgorithm.MatchingDictionary.Models.HLATypes;
using Nova.SearchAlgorithm.MatchingDictionary.Models.MatchingTypes;
using NUnit.Framework;

namespace Nova.SearchAlgorithm.Test.MatchingDictionary.Services.HlaMatching
{
    [TestFixtureSource(typeof(HlaMatchingTestFixtureArgs), "MatchedSerology")]
    [UseReporter(typeof(NUnitReporter))]
    public class SerologyToPGroupsMatchingTest : MatchedOnTestBase<IMatchedHla>
    {
        public SerologyToPGroupsMatchingTest(IEnumerable<IMatchedHla> matchingTypes) : base(matchingTypes)
        {
        }

        [Test]
        public void ValidSerologyHaveAtLeastOnePGroup()
        {
            var pGroupCounts = MatchingTypes
                .Where(m => !m.HlaType.IsDeleted && m.HlaType is Serology)
                .Select(m => new { m.HlaType, PGroupCount = m.MatchingPGroups.Count() })
                .ToList();

            Assert.IsTrue(pGroupCounts.All(p => p.PGroupCount > 0));
        }

        [Test]
        public void BroadWhereSplitHasAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("A", "9"));
        }

        [Test]
        public void BroadWhereSplitHasNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("DQB1", "1"));
        }

        [Test]
        public void BroadHasSplitAndAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "21"));
        }

        [Test]
        public void SplitHasAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "51"));
        }

        [Test]
        public void SplitNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("C", "10"));
        }

        [Test]
        public void AssociatedWithSplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "3902"));
        }

        [Test]
        public void AssociatedWithBroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "4005"));
        }

        [Test]
        public void AssociatedWithNotSplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("DRB1", "103"));
        }

        [Test]
        public void NotSplitWithAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("A", "2"));
        }

        [Test]
        public void NotSplitNoAssociatedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("A", "1"));
        }

        [Test]
        public void DeletedPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("C", "11"));
        }

        [Test]
        public void B15BroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "15"));
        }

        [Test]
        public void B15SplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "62"));
        }

        [Test]
        public void B70BroadPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "70"));
        }

        [Test]
        public void B70SplitPGroupsAreCorrect()
        {
            Approvals.Verify(GetPGroupsAsString("B", "72"));
        }

        [Test]
        public void B15And70ShareMatchingPGroups()
        {
            var b15 = GetSingleMatchingType("B", "15").MatchingPGroups;
            var b70 = GetSingleMatchingType("B", "70").MatchingPGroups;

            Approvals.Verify(string.Join("\r\n", b15.Intersect(b70).OrderBy(p => p)));
        }

        private string GetPGroupsAsString(string locus, string serologyName)
        {
            return string.Join("\r\n",
                GetSingleMatchingType(locus, serologyName)
                .MatchingPGroups
                .OrderBy(p => p));
        }
    }
}
