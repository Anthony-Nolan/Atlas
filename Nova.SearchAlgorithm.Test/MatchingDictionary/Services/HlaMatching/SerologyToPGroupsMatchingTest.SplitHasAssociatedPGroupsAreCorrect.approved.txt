15:23
51:01P
51:02P
51:03
51:04
51:05
51:06P
51:07P
51:08P
51:09P
51:10
51:100
51:101
51:102
51:103
51:104
51:105
51:106P
51:107
51:108
51:109
51:111
51:112
51:113
51:114
51:115
51:116
51:117
51:119
51:12
51:120
51:121
51:122
51:123
51:124
51:125
51:126
51:127
51:128
51:129
51:130
51:131
51:132
51:133
51:134
51:135
51:136
51:137
51:138
51:139
51:13P
51:14
51:140
51:141
51:143
51:144
51:145
51:146
51:147
51:148
51:15
51:150
51:151
51:152
51:153
51:154
51:155
51:156
51:157
51:158P
51:159
51:16
51:160
51:161
51:162
51:163
51:167
51:168
51:17
51:170
51:171
51:172
51:173Q
51:174P
51:175
51:176
51:177
51:179
51:18
51:180
51:181
51:182
51:183
51:185
51:186
51:187
51:188
51:189
51:19
51:190
51:191
51:192
51:194
51:195
51:196
51:197
51:198
51:199
51:20
51:200
51:201
51:202
51:203
51:204
51:205
51:206
51:207
51:208
51:209
51:21
51:210
51:211
51:212
51:213
51:214
51:215
51:216
51:217
51:218
51:22
51:220
51:221
51:222
51:223
51:225
51:226
51:227
51:228
51:23
51:24P
51:26
51:28
51:29
51:31
51:33
51:34
51:35
51:36
51:37
51:38
51:39
51:40
51:42
51:43
51:45
51:46
51:49
51:50
51:52
51:53
51:54
51:55
51:56P
51:57
51:58
51:59
51:60
51:61P
51:62
51:63P
51:64
51:65
51:66
51:67
51:68
51:69
51:70
51:71
51:72
51:73
51:74
51:75
51:76
51:77
51:78P
51:79
51:80
51:81
51:82
51:83
51:84
51:85
51:86
51:87
51:88
51:89
51:90
51:91
51:92P
51:93
51:94
51:95
51:96
51:97
51:99
53:06
58:08P